### PR TITLE
ssh network filter (23/24): add client transport

### PR DESCRIPTION
**Pull request by kralicky**

### CODE REVIEW

File: source/extensions/filters/network/ssh/client_transport.cc [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+  }
+  case FrameTags::RequestCommon: {
+    if (!upstream_is_direct_tcpip_) {
+      return sendMessageToConnection(extractFrameMessage(frame));
+    }
+    auto& sshFrame = static_cast<const SSHRequestCommonFrame&>(frame);
+    return sshFrame.message().visit(
+      [this](const wire::ChannelDataMsg& msg) -> absl::StatusOr<size_t> {
+        Envoy::Buffer::OwnedImpl buffer;
+        // NB: ChannelDataMsg data is length-prefixed, but don't write the length here
```
Comment (kenjenkins) [inline source/extensions/filters/network/ssh/client_transport.cc:109]: Would it be clearer to word this more like: "For a direct-tcpip channel, we want to forward only the raw data bytes (i.e. without any length prefix)." ?

File: source/extensions/filters/network/ssh/client_transport.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  absl::StatusOr<MiddlewareResult> interceptMessage(wire::Message& ssh_msg) override;
+
+  AuthStateSharedPtr downstream_state_;
+  std::unique_ptr<UpstreamUserAuthService> user_auth_svc_;
+  std::unique_ptr<UpstreamConnectionService> connection_svc_;
+  std::unique_ptr<PingExtensionHandler> ping_handler_;
+
+  std::map<std::string, UpstreamService*> services_;
+
+  bool channel_id_remap_enabled_{false};
```
Comment (kenjenkins) [inline source/extensions/filters/network/ssh/client_transport.h:58]: Does it make sense to have a separate member variable for this, or can we infer that remapping is enabled by checking `!channel_id_mappings_.empty()`?

File: source/extensions/filters/network/ssh/client_transport.cc [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    [](wire::UnimplementedMsg&) {
+      ENVOY_LOG(debug, "received UnimplementedMsg");
+      return absl::OkStatus();
+    },
+    [](wire::IgnoreMsg&) {
+      ENVOY_LOG(debug, "received IgnoreMsg");
+      return absl::OkStatus();
+    },
+    [](wire::DebugMsg&) {
+      ENVOY_LOG(debug, "received DebugMsg");
```
Comment (kenjenkins) [inline source/extensions/filters/network/ssh/client_transport.cc:172]: Should we also log the message from the DebugMsg here?

File: source/extensions/filters/network/ssh/client_transport.cc [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+  forward(std::move(msg), FrameTags{tags | EffectiveHeader});
+}
+
+absl::StatusOr<MiddlewareResult> SshClientTransport::interceptMessage(wire::Message& ssh_msg) {
+  return ssh_msg.visit(
+    [&](wire::ChannelOpenConfirmationMsg& msg) -> absl::StatusOr<MiddlewareResult> {
+      const auto& info = downstream_state_->handoff_info;
+      if (info.handoff_in_progress) {
+        channel_id_mappings_[info.channel_info->internal_upstream_channel_id()] = msg.sender_channel;
+        // channel is open, now request a pty
```
Comment (kenjenkins) [inline source/extensions/filters/network/ssh/client_transport.cc:211]: Should we make this pty request conditional based on whether `info.pty_info` is not `nullptr`? (What happens if the downstream hasn't requested a pty?)

File: source/extensions/filters/network/ssh/client_transport.cc [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+      // this message won't be dispatched to the upstream userauth service, so we need to handle a
+      // couple of post-auth-success actions that it would normally do
+
+      // set the upstream ext info if available
+      if (auto info = peerExtInfo(); info.has_value()) {
+        authState().upstream_ext_info = std::move(info);
+      }
+      // unregister the user auth service
+      unregisterHandler(user_auth_svc_.get());
```
Comment (kenjenkins) [inline source/extensions/filters/network/ssh/client_transport.cc:261]: I think I made a change to let the userauth service unregister itself upon receiving a UserAuthSuccessMsg: https://github.com/pomerium/envoy-custom/blob/bb0f27376dbbdf65d9691f84d16729cd417536e2/source/extensions/filters/network/ssh/service_userauth.cc#L358

But handling it here probably makes more sense, as this is what calls registerMessageHandlers() initially. Would you mind deleting that unregisterHandler() call from UpstreamUserAuthService?

File: test/extensions/filters/network/ssh/client_transport_test.cc [hunk 6]
```diff
@@ -0,0 +1,993 @@
...
+    authState->server_version = "SSH-2.0-Envoy";
+    authState->channel_mode = ChannelMode::Handoff;
+    authState->handoff_info.handoff_in_progress = true;
+    authState->handoff_info.channel_info = std::make_unique<SSHDownstreamChannelInfo>();
+    authState->handoff_info.channel_info->set_downstream_channel_id(100);
+    authState->handoff_info.channel_info->set_channel_type("session");
+    authState->handoff_info.channel_info->set_internal_upstream_channel_id(200);
+    authState->handoff_info.channel_info->set_initial_window_size(64 * wire::MaxPacketSize);
+    authState->handoff_info.channel_info->set_max_packet_size(wire::MaxPacketSize);
+    authState->handoff_info.pty_info = std::make_unique<SSHDownstreamPTYInfo>();
```
Comment (kenjenkins) [inline test/extensions/filters/network/ssh/client_transport_test.cc:138]: Can we also test the handoff when the pty_info is null?

File: test/extensions/filters/network/ssh/client_transport_test.cc [hunk 7]
```diff
@@ -0,0 +1,993 @@
...
+}
+
+TEST_F(ClientTransportTest, UserAuthServiceRequest) {
+  StartTransportNormal();
+
+  ASSERT_OK(ExchangeExtInfo());
+  ASSERT_OK(HandleUserAuth());
+}
+
+TEST_F(ClientTransportTest, OpenChannelFromDownstream) {
```
Comment (kenjenkins) [inline test/extensions/filters/network/ssh/client_transport_test.cc:399]: Would it make sense to also test opening multiple channels here?

File: source/extensions/filters/network/ssh/client_transport.h (hunk 2 above)
Comment (kralicky) [reply source/extensions/filters/network/ssh/client_transport.h:58]: Since this has to be checked on every message, comparing a single bool here is probably quicker.

File: source/extensions/filters/network/ssh/client_transport.cc (hunk 5 above)
Comment (kralicky) [reply source/extensions/filters/network/ssh/client_transport.cc:261]: This is a special case - the logic here is only for handoff. Since we are intercepting it here, we have to handle the cleanup steps ourselves (see the comment above). In the normal case, it is fine for the userauth service to unregister itself.

Comment (kenjenkins) [reply source/extensions/filters/network/ssh/client_transport.cc:261]: Ah OK, sounds good.
